### PR TITLE
Streamline and tighten lintr checks

### LINKED ISSUE #218: Revisit lintr usage
Hi,

I would like to suggest streamlining and tightening automated checks performed by lintr in this package:

- The current workflow will lint all files, which results in annoying annotations on files you didn't touch and don't intend to touch in your PR. To remedy this, the lintr team has released a [`lint-changed-files.yaml` workflow](https://github.com/r-lib/actions/blob/7434f9e8cb77ffadb7bd5227f5393307fd516f77/examples/lint-changed-files.yaml), which will extract modified files and run lintr only on these files.
- You only get a tiny fraction of lintr's power by using only the default linters. lintr has [many](https://lintr.r-lib.org/reference/best_practices_linters.html), [many](https://lintr.r-lib.org/reference/efficiency_linters.html) more linters that are turned off by default and that can identify potential bugs or inefficient code. Default linters are not based on stability or utility but only on whether they encode the tidyverse style or not. An immediate consequence of this is that only style annotations will be provided and you will miss on all the more advanced static analysis.

 If you believe a change here could be helpful, I can submit a PR to:

- Switch to the `lint-changed-files.yaml` workflow
- Try to turn on all linters
- Fix lints in the current codebase and selectively turn off linters that are identified as incompatible with some of your design choice / coding style.

### PR DESCRIPTION
**Pull request by Bisaloo**
Fix #218 

- 325212a replaces the workflow to only run in changed files
- c7ecfb7 modifies `.lintr` configuration to enable non-default linters
- 84edd71 to de657ed fix lints from one specific linter at a time
- 179a4de modifies `.lintr` configuration to disable linters that cause too many false positives or clash with design choices made in this repo (e.g. indents in function argument definition)

If all goes well, only one lint should be remaining. It flags the fact that the [`no_contrasts` variable](https://github.com/epinowcast/epinowcast/blob/main/R/formula-tools.R#L630-L632) is not used `enw_formula()`. I'm not completely sure what the right fix is here. Whether `no_contrasts` should just be removed here or if the wrong variable is passed at some point in place of `no_contrasts`.

The other changes should mostly result in a more coherent styling but also in a slightly more performant code in some cases (5d7595f, d2e57ea, 4ae9ed2, 2906dc3 among others).

### CODE REVIEW

**Comment by pearsonca** [comment]
General question to correct my own ignorance: is there a way for me to run *this* check locally? As in, before committing/pushing, a one-line-at-prompt, use this configuration to lint?

**Comment by Bisaloo** [comment]
Very interesting question @pearsonca! I've never really considered it. 

I usually just run `lintr::lint_package()`, which will automatically pick up the `.lintr` config file. But if you want to run this specifically on changed file, there must be a way to pipe the output from `git diff` (or from one of the git wrapper R package).

**Comment by Bisaloo** [comment]
Not one line but the code from the GHA can be modified to work locally on changed files, before you `add` them.

```r
changed_files <- gert::git_diff()$new
all_files <- list.files(recursive = TRUE)
exclusions_list <- as.list(setdiff(all_files, changed_files))
lintr::lint_package(exclusions = exclusions_list)
```

**Comment by Bisaloo** [comment]
I again forgot to base my changed on `develop` :cry:

**Comment by seabbs** [comment]
> I again forgot to base my changes on develop 😢

This is super easy to do. Any thoughts on ways we can make our develop approach clearer here? We could also just drop it because we now have control releases to `r-universe`. Hmm will add this to the docket for the next community call.

**Comment by pearsonca** [comment]
re basing off `develop`, is there a way to set the default base on the make-a-branch-from-this-issue clicky bits?

**Comment by seabbs** [comment]
> re basing off develop, is there a way to set the default base on the make-a-branch-from-this-issue clicky bits?

I have ineffectually looked for this a few times as it seems like there must be.

File: R/check.R
Comment (seabbs) [inline R/check.R:20]: So many nice `base` functions coming out you play ❤️

**Review by seabbs — CHANGES_REQUESTED** [review]
This all looks good to me and is a  lot of work so thanks again @Bisaloo. The only outstanding actions we need before we can go ahead is to iterate the development version, and add a news item linking to this PR. We should also open an issue to address the `no_contrasts` linting warning but I am happy to do that.

@pearsonca or @adrian-lison (or anyone else @epinowcast/developers ) who has a chance it would be great to get a second pair of eyes on this just to check nothing has slipped through

File: .lintr
Comment (Bisaloo) [reply .lintr:2]: No, I think you're fine. I personally don't really like most of the disabled linters here because I find them too noisy.

- `implicit_integer_linter` requires using `1L` instead of `1`, which becomes annoying when you have `df[1, ]`
- `extraction_operator_linter` requires using `[[` over `[` whenever possible, which I find less readable
- `undesirable_function_linter` flags `library()` use in vignettes and other files outside of `R/` so too many false positive. It could be reactivated with a subset of functions you want to avoid everywhere though (e.g. `sapply()`
- `function_argument_linter` requires that arguments with defaults appear after arguments without defaults. It may be something to consider here as this is part of the official tidyverse guidance. But I don't always follow this rule myself.
- `function_argument_linter` is still experimental and has too many false positives. It also clashes with your formatting style for argument definition.
- `object_name_linter` flags things like `is.Date()`

**Comment by Bisaloo** [comment]
>  to iterate the development version

Making sure I understand: should I increment to package version number to 0.2.0.7000?

What is the preferred way to resolve conflicts? Should I rebase locally and force push or should I create a merge commit here?

**Comment by seabbs** [comment]
> Making sure I understand: should I increment to package version number to 0.2.0.7000?

(yes though now it is at .8000 - I read somewhere this was a good idea and we decided to try it out but it is a real pain)

What is the preferred way to resolve conflicts? Should I rebase locally and force push or should I create a merge commit here?

Whatever you would like. As we are using squash merging (😱) I am generally not hugely worried about PR commits as long as they are clear.

File: .lintr
Comment (seabbs) [reply .lintr:2]: Okay, all good to know. I hadn't realised arguments with defaults ahead of args without was good practice. We can probably address this on a case-by-case basis but as you say there will likely be some areas where we still want to do this.

> formatting style for argument definition.

By this do you mean having 2nd line (and so on indented to match the first line? I am not particularly wedded to that approach. 

We could potentially consider adding a subset of this info as comments the `.lintr` for future updating ease etc. but not required.

Comment (Bisaloo) [reply .lintr:2]: > By this do you mean having 2nd line (and so on indented to match the first line? I am not particularly wedded to that approach.

Yes.

An alternative could be:

```r
f <- function(
  arg1,
  arg2
)
```

But as mentioned, even with this, it may be more reasonable to keep `indentation_linter` off for now because many false positives have been reported.

> We could potentially consider adding a subset of this info as comments the .lintr for future updating ease etc. but not required.

lintr documentation is quite good on this aspect so at best it would be redundant and at worst it could become outdated so I would probably not recommend it. But if it's something you'd like to have, I can still add it :shrug:

Comment (seabbs) [reply .lintr:2]: > lintr documentation is quite good on this aspect so at best it would be redundant and at worst it could become outdated so I would probably not recommend it.

No this makes sense. Good point. 

> An alternative could be:

Ah right okay. I do quite like this approach but it can lead to very long files.

**Review by seabbs — COMMENTED** [review]
This looks good to go to me. Happy to see this merged in when you are ready @Bisaloo. Thanks again for the nice clean PR.

**Comment by Bisaloo** [comment]
> Happy to see this merged in when you are ready @Bisaloo.

I think it's good to go. I can't merge it myself as one check is failing but feel free to do it whenever convenient.

**Review by seabbs — APPROVED** [review]
This still all looks good to me so going to merge without a second review. Note this will cause lint checks on develop to fail but we have an issue open to resolve.

**Pull request merged**
